src/pruner.test.ts

```diff
@@ -0,0 +1,518 @@
+    it('should not include object', () => {
+      const parameters = {
+        prop1: {
+          prop2: 'something',
+          prop3: 42
+        }
+      }
+      const schema: JSONSchema7 = {
+        type: 'object',
+        properties: {
+          prop1: {
+            type: 'object',
+            properties: {
+              prop2: {
+                type: 'string',
+                default: 'something'
+              },
+              prop3: {
+                type: 'number'
+              }
+            }
+          }
+        },
+        additionalProperties: false
+      }
+
+      const result = pruneDefaults(parameters, schema)
+
+      const expected = {
+        prop1: {
+          prop3: 42
+        }
+      }
+      expect(result).toEqual(expected)
+    })
+  })
+})
+
+describe('given array items is string and has default', () => {
+  it('should not include prop', () => {
+    const parameters = {
+      prop1: ['something']
+    }
+    const schema: JSONSchema7 = {
+      type: 'object',
+      properties: {
+        prop1: {
+          type: 'array',
+          items: {
+            type: 'string',
+            default: 'something'
+          }
+        }
+      },
+      additionalProperties: false
+    }
+
+    const result = pruneDefaults(parameters, schema)
+
+    const expected = {}
+    expect(result).toEqual(expected)
+  })
+})
+
+describe('given array items is string and has item equal to default and item which is non-default', () => {
+  it('should not include prop', () => {
+    const parameters = {
+      prop1: ['something', 'somethingelse']
+    }
+    const schema: JSONSchema7 = {
+      type: 'object',
+      properties: {
+        prop1: {
+          type: 'array',
+          items: {
+            type: 'string',
+            default: 'something'
+          }
+        }
+      },
+      additionalProperties: false
+    }
+
+    const result = pruneDefaults(parameters, schema)
+
+    const expected = {
+      prop1: ['somethingelse']
+    }
+    expect(result).toEqual(expected)
+  })
+})
+
+describe('given array items is object and has nested prop equal to default and another item with same prop with non-default value', () => {
+  it('should not include prop', () => {
+    const parameters = {
+      prop1: [{ prop2: 'something' }, { prop2: 'somethingelse' }]
+    }
+    const schema: JSONSchema7 = {
+      type: 'object',
+      properties: {
+        prop1: {
+          type: 'array',
+          items: {
+            type: 'object',
+            properties: {
+              prop2: {
+                type: 'string',
+                default: 'something'
+              }
+            },
+            additionalProperties: false
+          }
+        }
+      },
+      additionalProperties: false
+    }
+
+    const result = pruneDefaults(parameters, schema)
+
+    const expected = {
+      prop1: [{ prop2: 'somethingelse' }]
```

Comment:
Not sure I understand this test case. When would you ever get to see a duplicate prop? 

The current test assumes the second one overwrites the first, but I don't think this is even possible through the workflow builder, is it? 

The alternative I can think of is that, say `prop1` is a function that takes `prop2` as an input. Now, if you specify multiple values for `prop2`, a reasonable expectation could be that the function will be executed for all instances of prop2. In that case, you want to conserve the length of the array.

Reply:
There are parameters where the order/length of the array does not matter, because the index of the item is not linked to another like the index of a molecule. So the current code works for those parameters, but you are correct for parameters where the index matters it breaks. 

I do not have a way to tell if the index matters so it is safer to keep default array items to not change the indices after it.
I tried to fix it but ran into downstream problems so spinned off #55